KT-48594: Preserve symlinks during framework copy

rickclephas opened the pull request:
The Gradle `Copy` tasks which are currently used to copy the Xcode frameworks erase the symlinks within those frameworks.
This results in failing macOS app build (see [KT-48594](https://youtrack.jetbrains.com/issue/KT-48594)).

This can be fixed by using the `cp -R` command, which doesn't follow symlinks but copies them instead.
Though `cp` doesn't support copying symlinks that already exists, so `rm -r` is used to make sure the destination files don't exist.

terrakok commented:
Hi! Thank you for the fix. Could you add test for a scenario with macos framework embedding to `libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/AppleFrameworkIT.kt`?

terrakok commented on libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/apple/AppleXcodeTasks.kt:
```diff
@@ -200,4 +202,15 @@ private val Framework.namePrefix: String
\ No newline at end of file
+    buildDir.resolve("xcode-frameworks").resolve(frameworkSearchDir)
+
+private fun Project.copyFiles(files: Array<File>, destDir: File) {
```
It would be perfect to add comment about symlinks behavior with YouTrack issue ID.

rickclephas commented:
@terrakok I have added the `macosX64` target to the test project and a test to verify that he macOS framework contains a symlink. Also added a comment with link to the YouTrack issue.

terrakok commented on libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/AppleFrameworkIT.kt:
```diff
@@ -78,6 +78,27 @@ class AppleFrameworkIT : BaseGradleIT() {
+                assertSuccessful()
+                assertTasksExecuted(":shared:assembleDebugAppleFrameworkForXcodeMacosX64")
+                assertFileExists("/shared/build/xcode-frameworks/debug/macosx/sdk.framework/Headers")
+                assertFileIsSymlink("/shared/build/xcode-frameworks/debug/macosx/sdk.framework/Headers")
```
This test checks success build but we need to check macos framework copying.
You can do it this way:
```
customEnvironmentVariables = mapOf(
  "CONFIGURATION" to "debug",
  "SDK_NAME" to "macosx",
  "ARCHS" to "x86_64",
  "TARGET_BUILD_DIR" to workingDir.absolutePath,
  "FRAMEWORKS_FOLDER_PATH" to "build/xcode-derived" //<-- path to embedding assembled framework
)
build(":shared:embedAndSignAppleFrameworkForXcode", options = options) {
  assertSuccessful()
  assertTasksExecuted(":shared:assembleDebugAppleFrameworkForXcodeMacosX64")
  assertFileExists("/build/xcode-derived/sdk.framework/Headers")
  assertFileIsSymlink("/build/xcode-derived/sdk.framework/Headers")
}
```

rickclephas replied:
Alright will check this later today. The current test verifies the first copy step to the `xcode-frameworks` folder which was already missing the symlinks.

rickclephas replied:
Updated so it's now verifying that the symlinks exists in both the `build/xcode-frameworks` folder and the Xcode derived data folder.

terrakok commented on libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/resources/testProject/sharedAppleFramework/shared/build.gradle.kts:
```diff
@@ -22,6 +23,9 @@ kotlin {
         val commonMain by getting
 
+        val macosX64Main by getting {
+            dependsOn(commonMain)
```
it is unnecessary. all targets depend on common by default

terrakok commented:
When you changed gradle `Copy` task to `DefaultTask` up-to-date checks were missed.
I think we should create abstract `FrameworkCopy` task with correct inputs and outputs.

And one more moment, there are not `cp` and `rm` commands on Windows. So, in the new task we should check host OS with error message before copying.

rickclephas commented:
> When you changed gradle `Copy` task to `DefaultTask` up-to-date checks were missed. I think we should create abstract `FrameworkCopy` task with correct inputs and outputs.

Nice catch! Didn't think about that 😅. I have moved the copy part into a `FrameworkCopy` task which makes it cacheable.

> And one more moment, there are not `cp` and `rm` commands on Windows. So, in the new task we should check host OS with error message before copying.

This shouldn't be a problem since the copy tasks will never be executed on non macOS hosts:
https://github.com/JetBrains/kotlin/blob/650e1f65adea439426130a4b549df9a9c7493784/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/apple/AppleXcodeTasks.kt#L77

terrakok commented on libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/apple/AppleXcodeTasks.kt:
```diff
@@ -200,4 +203,32 @@ private val Framework.namePrefix: String
+private abstract class FrameworkCopy : DefaultTask() {
+
+    @get:InputFiles
+    var files: FileCollection? = null
```
We need little bit more checks :) and since the task is abstract we can avoid nullability
```
@get:PathSensitive(PathSensitivity.ABSOLUTE)
@get:IgnoreEmptyDirectories
@get:InputFiles
@get:SkipWhenEmpty
abstract var files: FileCollection

@get:OutputDirectory
abstract var destDir: File
```

terrakok commented on libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/apple/AppleXcodeTasks.kt:
```diff
@@ -176,11 +180,10 @@ internal fun Project.registerEmbedAndSignAppleFrameworkTask(framework: Framework
-            task.doLast {
+        task.files = files(File(appleFrameworkDir(envFrameworkSearchDir), framework.outputFile.name))
+        task.destDir = envEmbeddedFrameworksDir
+        task.doLast {
+            if (envSign != null) {
```
i think better:
```
if (envSign != null) task.doLast {
```

terrakok commented:
I think it is almost ready for merging, so let's squash commits into one please. i'll be rebase it on master after all checks are green

terrakok commented:
merged to master: https://github.com/JetBrains/kotlin/commit/2c8086bbfe48b9cfb62e373205c5032fd3731cc3
